tests/device_tests/test_pin.py

```diff
@@ -0,0 +1,123 @@
+# This file is part of the Trezor project.
+#
+# Copyright (C) 2012-2019 SatoshiLabs and contributors
+#
+# This library is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License version 3
+# as published by the Free Software Foundation.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the License along with this library.
+# If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
+
+import time
+
+import pytest
+
+from trezorlib import messages
+from trezorlib.exceptions import PinException
+
+from ..common import get_test_address
+
+PIN4 = "1234"
+BAD_PIN = "5678"
+
+pytestmark = pytest.mark.setup_client(pin=PIN4)
+
+# FIXME TODO Add passphrase tests
+
+
+@pytest.mark.setup_client(pin=None)
+def test_no_protection(client):
+    with client:
+        client.set_expected_responses([messages.Address])
+        get_test_address(client)
+
+
+def test_correct_pin(client):
+    with client:
+        client.use_pin_sequence([PIN4])
+        # Expected responses differ between T1 and TT
+        is_t1 = client.features.model == "1"
+        client.set_expected_responses(
+            [
+                (is_t1, messages.PinMatrixRequest),
+                (
+                    not is_t1,
+                    messages.ButtonRequest(code=messages.ButtonRequestType.PinEntry),
+                ),
+                messages.Address,
+            ]
+        )
+        # client.set_expected_responses([messages.ButtonRequest, messages.Address])
+        get_test_address(client)
+
+
+@pytest.mark.skip_t2
+def test_incorrect_pin_t1(client):
+    with pytest.raises(PinException):
+        client.use_pin_sequence([BAD_PIN])
+        get_test_address(client)
+
+
+@pytest.mark.skip_t1
+def test_incorrect_pin_t2(client):
+    with client:
+        # After first incorrect attempt, TT will not raise an error, but instead ask for another attempt
+        client.use_pin_sequence([BAD_PIN, PIN4])
+        client.set_expected_responses(
+            [
+                messages.ButtonRequest(code=messages.ButtonRequestType.PinEntry),
+                messages.ButtonRequest(code=messages.ButtonRequestType.PinEntry),
+                messages.Address,
+            ]
+        )
+        get_test_address(client)
+
+
+def _test_backoff(attempts, start):
+    """Helper to assert the exponentially growing delay after incorrect PIN attempts"""
+    if attempts <= 1:
+        expected = 0
+    else:
+        expected = (2 ** (attempts - 1)) - 1
+    got = round(time.time() - start, 2)
+    assert got >= expected
+
+
+@pytest.mark.skip_t2
+def test_exponential_backoff_t1(client):
+    for attempt in range(1, 4):
```

Comment:
```suggestion
    for attempt in range(3):
```

together with the above suggestion the results will be the same

Reply:
Fixed in 699ae28